tools/clang/lib/SPIRV/DeclResultIdMapper.cpp

```diff
@@ -574,6 +574,20 @@ SpirvInstruction *DeclResultIdMapper::getDeclEvalInfo(const ValueDecl *decl,
           {spvBuilder.getConstantInt(
               astContext.IntTy, llvm::APInt(32, info->indexInCTBuffer, true))},
           loc);
+    } else if (auto *type = info->instr->getResultType()) {
+      const auto *ptrTy = dyn_cast<HybridPointerType>(type);
+
+      // If it is a local variable or function parameter with a bindless
+      // array of an opaque type, we have to load it because we pass a
+      // pointer of a global variable that has the bindless opaque array.
+      if (ptrTy != nullptr && isOpaqueArrayType(decl->getType()) &&
+          !decl->getType()->isConstantArrayType()) {
+        auto *load = spvBuilder.createLoad(ptrTy, info->instr, loc);
+        load->setRValue(false);
```

Comment:
Yes, I think we have to use it as a Lvalue.

In the function definition like
```
float4 SampleArray(Texture2D src[], uint index, float2 uv) {
  return src[index].Sample(DummySampler, uv);
}
```

it accesses to `Texture2D src[]`. Since `src[index]` has `ArraySubscriptExpr`, we want to handle it using `OpAccessChain` or `OpLoad`/`OpStore` to get a pointer of `src`. If we set it as a Rvalue, we have to interpret "it cannot be a Lvalue" and we have to create a variable to get the pointer.

Therefore, as far as I understand setting it as "Rvalue" has a clear deficiency:
- We have to generate some extra code to access it e.g., temporary variable generation in `SpirvEmitter::turnIntoElementPtr()`. Actually the current `SpirvEmitter::turnIntoElementPtr()` generates invalid result type for `src[index]` which cannot be fixed by legalization. I guess there might be more corner cases.

See the following code that DXC generates when setting it Rvalue:
```
 100 %SampleArray = OpFunction %v4float None %47
 101         %src = OpFunctionParameter %_ptr_Function__ptr_UniformConstant__runtimearr_type_2d_image
 102       %index = OpFunctionParameter %_ptr_Function_uint
 103          %uv = OpFunctionParameter %_ptr_Function_v2float
 104  %bb_entry_0 = OpLabel
 105   %temp_var_ = OpVariable %_ptr_Function__runtimearr_type_2d_image Function
 106          %54 = OpLoad %uint %index
 107          %55 = OpLoad %_ptr_UniformConstant__runtimearr_type_2d_image %src
 108                OpStore %temp_var_ %55
 109          %57 = OpAccessChain %_ptr_Function_type_2d_image %temp_var_ %54
 110          %58 = OpLoad %type_2d_image %57
...
```
In line `108`, it is not only inefficient but also has a bug that saves `%_ptr_UniformConstant__runtimearr_type_2d_image` type `%55` into `%temp_var_` (type mismatch).

So .. for storing it into some variables, Rvalue might be fine but for reading it, Lvalue looks like a correct choice.

Reply:
Thanks for the explanation. SG.